keps/sig-api-machinery/4153-declarative-validation/README.md

```diff
@@ -0,0 +1,1793 @@
+| ---------------------- | ---------------------------------------------------------------- | ---------------------------------------------------------------------------------- | ---------------- |
+| string format          | `+format={format name}`                                          | `format`                                                                           | Invalid          |
+| size limits            | `+min{Length,Properties,Items}`, `+max{Length,Properties,Items}` | `min{Length,Properties,Items}`, `max{Length,Properties,Items}`                     | TooMany, TooLong |
+| numeric limits         | `+minimum`, `+maximum`, `+exclusiveMinimum`, `+exclusiveMaximum` | `minimum`, `maximum`, `exclusiveMinimum`, `exclusiveMaximum`                       | Invalid          |
+| required fields        | `+optional` (exists today)                                       | `required`                                                                         | Required         |
+| enum values            | `+enum` (exists today)                                           | `enum`                                                                             | NotSupported     |
+| uniqueness             | `listType=set` (sets and map keys)                               | `x-kubernetes-list-type`                                                           | Invalid          |
+| regex matches          | `+pattern`                                                       | `pattern`                                                                          | Invalid          |
+| cross field validation | `validationRule="{CEL expression}"`                              | `x-kubernetes-validations`                                                         | user specified   |
+| [transition rules][1]  | `validationRule="{CEL expression using oldSelf}"`                | `x-kubernetes-validations`                                                         | user specified   |
+| special case: name format | `+nameFormat={format name}`                                   | `x-kubernetes-validations` (see "format" section below for details)                | Invalid          |
+
+xref: [Kubernetes mapping of OpenAPI value validations to response code](https://github.com/kubernetes/kubernetes/blob/7adcb3cb37084750ab572a179ba01b9878b8de5f/staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/validation/validation.go#L60)
+
+### Unions
+
+Unions are informally defined but common in built-in types.
+
+We will convert most union validation to CEL expressions. For example:
+
+```go
+// staging/src/k8s.io/api/apps/v1/types.go
+
+type DeploymentSpec struct {
+  // ...
+  //
+  // +minimum=0
+  Replicas string `json...`
+  
+  // ...
+  // 
+  // +validationRule="!(self.type == 'Recreate') || !has(self.rollingUpdate)",message="may not be specified when strategy `type` is Recreate",reason=Forbidden,field="rollingUpdate"
```

Comment:
The HTTP code might be Forbidden, but I think we should craft a better message for this too-common case.

e.g.

```
// +contextual="self.rollingUpdate",iff="self.type == 'RollingUpdate')
```

Something that makes it really clear - "self.rollingUpdate" is only valid if "self.type == 'RollingUpdate', and produces a consistent error like "Forbidden: spec.rollingUpdate may only be specified when spec.type is RollingUpdate"

I don't like my proposed syntax, but you get the idea, I hope.

Reply:
Added this to the note about common idioms.